custom_components/sonnenbatterie/sensor.py

```diff
@@ -31,6 +31,123 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True, kw_only=True)
+class SonnenbatterieSensorEntityDescription(SensorEntityDescription):
+    """Describes Example sensor entity."""
+
+    exists_fn: Callable[[SonnenBatterieCoordinator], bool] = lambda _: True
+    value_fn: Callable[[SonnenBatterieCoordinator], StateType]
+
+
+SENSORS: tuple[SonnenbatterieSensorEntityDescription, ...] = (
```

Comment:
Ich finde den Ansatz über die `SensorEntityDescription` total gut, befürchte aber, dass sich das so (also mit einer Konstanten) nicht durchhalten lassen wird. Spätestens wenn es an die dynamisch erzeugten Sensoren (z.B. "in/out" - also solche ohne _direkten_ Bezug zu einem von der API gelieferten Wert) geht oder die Funktionen zum Text-Mapping scheint mir die Definition über eine Konstante zumindest auf den ersten Blick nicht flexibel genug.

Wollte das nur schnell anmerken, da die entsprechenden Definitionen erst weiter unten in der `mappings.py` kommen.

Reply:
Da hast du recht, da wird es schwierig. Ich habe aber daran gedacht, die Werte beim update im coordinator bereits vorzurechnen, damit man sie dann einfach in der value function aufrufen kann. Damit die Liste nicht riesen groß wird, muss ich mal schauen, ob ich die entities auch dynamisch erzeugt zur `SENSORS` Liste hinzufügen kann.
Das werde ich aber wahrscheinlich erst später machen und vorerst jede entity manuell eintragen.

Reply:
Man sieht ja schon jetzt mit der ausgelagerten Sensor-Definition, was für eine Menge an Getippe das wird. Spontan sehe ich zwei Probleme:
1. Die diversen Sonnenbatterien liefern unterschiedliche API-Antworten. Das hatte ich versucht über ˋaliasˋ abzubilden/abzufangen, um jeweils einen global gültigen Wert für alle Inkarnationen der API bereitstellen zu können. Das scheint mir so bei deinem Weg gar nicht mehr möglich zu sein.
2. Durch die Verlagerung vom Programmatischen zum Deklarativen muss bei deiner Variante jede gewünschte Option als eigene Entity definiert werden, und dazu noch jeweils eigene, aber eigentlich immer identische ˋvalue_fnˋ und ˋexist_fnˋ. Mein Ansatz war da eher, den Anwendern mehr Flexibilität zu geben und das über "Flags" zu steuern. Was ja in deinem Ansatz prinzipiell nicht ausgeschlossen ist, nur denke ich, dass die Basisstruktur etwas zu starr ausgelegt ist. 

Dynamisches erzeugen der Sensorliste sollte nicht allzu schwierig sein, man braucht ja nicht zwingend ein ˋtupleˋ dafür verwenden.

Reply:
Da hast du schon recht, ich bin auch noch nicht so ganz zufrieden damit. Ich habe in den letzten Änderungen die powermeter dynamische erstellt, was ganz gut funktioniert hat. Vielleicht nutze ich auch doch wieder die Mappings und nutze sie, um die anderen Entities in der Liste auch dynamisch zu erzeugen.

Reply:
> Die diversen Sonnenbatterien liefern unterschiedliche API-Antworten. Das hatte ich versucht über ˋaliasˋ abzubilden/abzufangen, um jeweils einen global gültigen Wert für alle Inkarnationen der API bereitstellen zu können. Das scheint mir so bei deinem Weg gar nicht mehr möglich zu sein.

Kannst du mir dazu ein Beispiel geben? Ich kann dazu keine Option finden. Meinst du vielleicht die `aka` Option? So wie ich das verstehe erstellt die aber 2 gleiche Sensoren mit unterschiedlichen Namen und nutzt die exakt gleiche API Werte.
Aber wenn es wirklich unterschiedliche API Endpunkte je nach Sonnenbatterie gibt könnte man in value_fn auch einfach ein `or` verwenden.

Reply:
Korrekt, ich meinte die `aka` Option, du hast sie nur falsch rum verstanden ;) Wenn du zum Beispiel auf den Aka-Sensor `state_netfrequency` schaust, dann wird der _entweder_ aus `battery_system/battery_system_fac` _oder_ aus `inverter/status/status/fac` _oder_ aus `inverter/status/fac` generiert (whichever comes first ...) weil der Wert eben bei unterschiedlichen SonnenBatterien an diesen unterschiedlichen Stellen zu finden ist. Gerade bei `inverter/status/status/fac` wird das mit `or` schwierig, weil du da nicht `None` zurück kriegst, sondern einen Fehler, wenn das gesamte zweite `status` schon nicht vorhanden ist, da du keinen named index von `None` machen kannst.

Reply:
> Gerade bei inverter/status/status/fac wird das mit or schwierig, weil du da nicht None zurück kriegst, sondern einen Fehler, wenn das gesamte zweite status schon nicht vorhanden ist, da du keinen named index von None machen kannst.

Das ist nicht wirklich ein Problem, anstelle von ['wert'] kann man die `.get('wert', {})` Funktion zu verwenden, bei der der zweite Parameter der Standardwert ist und sich somit der Fehler vermeiden lässt. So wurde es auch in anderen Integrationen gelöst (z.B. roomba).